src/workers/AGENTS.md

```diff
@@ -0,0 +1,380 @@
+    topic: 'my-new-topic', // Topic must exist in the "streams" repository
+    subscription: 'my-new-subscription',
+    args: {
+      // Optional: configure ack deadline, dead letter policy, etc.
+      ackDeadlineSeconds: 60,
+    },
+  },
+];
+```
+
+### Step 5: Publish Messages
+
+Use `triggerTypedEvent` to publish messages (this is the standard approach):
+
+```typescript
+import { triggerTypedEvent } from '../common/typedPubsub';
+
+// In your resolver or service:
+await triggerTypedEvent(logger, 'my-new-topic', {
+  userId: '123',
+  action: 'created',
+});
+```
+
+## Change Data Capture (CDC)
+
+CDC allows us to react to database changes without distributed transactions. Debezium captures PostgreSQL changes and publishes them to Pub/Sub.
+
+### How CDC Works
+
+1. **Debezium** monitors PostgreSQL WAL (Write-Ahead Log)
+2. Changes are published to Pub/Sub topics
+3. Workers subscribe to CDC topics and react to changes
+4. This enables eventual consistency across systems
+
+### CDC Worker Example
+
+```typescript
+import { Worker, messageToJson } from './worker';
+import { ChangeMessage } from '../types';
+
+const worker: Worker = {
+  subscription: 'api-cdc',
+  maxMessages: 20, // Process multiple messages at once
+  handler: async (message, con, logger): Promise<void> => {
+    const data: ChangeMessage<any> = messageToJson(message);
+    
+    // Skip heartbeat and read operations
+    if (
+      data.schema.name === 'io.debezium.connector.common.Heartbeat' ||
+      data.payload.op === 'r'
+    ) {
+      return;
+    }
+    
+    // React to specific table changes
+    switch (data.payload.source.table) {
+      case 'user':
+        await handleUserChange(con, logger, data);
+        break;
+      case 'post':
+        await handlePostChange(con, logger, data);
+        break;
+    }
+  },
+};
+```
+
+### CDC Tables
+
+CDC is configured for specific tables in `.infra/application.properties`. Common patterns:
+
+- React to user changes for external sync (CIO, etc.)
+- React to post changes for indexing, notifications
+- React to vote changes for reputation calculations
+- React to comment changes for notifications
+
+## Worker Execution
+
+### Background Processor
+
+Workers run in the background processor:
+
+```bash
+pnpm run dev:background
+```
+
+This starts `src/background.ts`, which:
+1. Connects to the database
+2. Initializes Pub/Sub client
+3. Subscribes all registered workers to their subscriptions
+4. Processes messages as they arrive
+
+### Message Flow
+
+1. **Publish**: Application code publishes a message to a Pub/Sub topic
+2. **Subscribe**: Worker subscribes to the subscription
+3. **Receive**: Pub/Sub delivers the message to the worker
+4. **Process**: Worker handler executes
+5. **Ack/Nack**: Worker acknowledges (ack) on success or negative-acknowledges (nack) on failure
+   - **Ack**: Message is removed from subscription
+   - **Nack**: Message is redelivered (with exponential backoff)
+
+### Error Handling
```

Comment:
Something it's also good it validated params.
EG it's userId it could be user deleted in meantime so it should query internally to validate preferably?

Reply:
In some cases for instance now with external messages (gondul example where match can be made, but used deleted account in mean time)

The worker ideally should query user from database vs simply hoping it exists to avoid endless errors.